### ISSUE TITLE
Q: How to print way or result userdata with pprint.lua?

### ISSUE THREAD

**Issue by mloskot** [Project-OSRM/osrm-backend#3718]
While working with Lua profiles, I found that I can get `print` output from inside my `node_function` or `way_function`. That is not only when I pass a profile through the `profiles/lib/profile_debugger.lua`, but even if profile is executed by `osrm-extract`.

For example, I tried Lua `print`:
```
function way_function (way, result)
  print('way.id = ', way:id())
end
```
Now, if I run `osrm-extract -p my.lua my.osm`, I see `way.id = 1` printed in the console. _That is useful_.

The `pprint` function from `profiles/lib/pprint.lua` does the job well too, for basic types...

-----

Next, I tried to dump the `way` and `result` objects with `pprint`

* I put this into my profile script:
```
pprint(way)
pprint(result)
```

* I overwrite the default `pprint` option `show_userdata = false` with `pprint.setup({show_userdata = true})` or  (equivalent?) `pprint.defaults.show_userdata = true`.

* I run `osrm-extract` with my profile and I see output like this:

```
[[userdata 1]]
[[userdata 1]]
```

Both, `way` and `result` have Lua type of `userdata` type.

I'm not Lua user or expert whatsoever, touching Lua for the first time, but it looks `pprint` is not really able to inspect and dump the actual `userdata` content, is it?

Is that what the `make_fixed_formatter` defined in `profiles/lib/pprint.lua` supposed to output?

Or, is there any magic I can use to make `pprint` work with the `userdata` objects?

**Comment by emiltin** [comment]
When you run osrm-extract, way and result are C++ objects being passed to LUA from the OSRM engine, which is why they can't be printed easily. What you can try is to add a LUA helper function to dump them, eg:

```lua
function print_way(way)
  print(way.forward_mode)
  print(way.backward_mode)
end
```

Another thing you can do is to make a copy of the file debug_example.lua and edit it to set some input tags. When you run it (see instructions in the file), it will call your profile code, passing a LUA struct as the way, so in this case you'll be able to pprint the way. The debug_example.lua file will show the output from the way function.

**Comment by mloskot** [comment]
@emiltin 
> When you run osrm-extract, way and result are C++ objects being passed to LUA from the OSRM engine, which is why they can't be printed easily.

Right. Looks like they're types delivered by libosmium.

> `print(way.forward_mode)`

Yes, that is indeed what I ended up with. I just thought `pprint` would dump all members automagically and include all of them.

> Another thing you can do is to make a copy of the file debug_example.lua (...)

Yes, I learned about the debugging utilities too. It's handy.

BTW, any chance to comment on my [Is profile_debugger.lua out of sync?](https://lists.openstreetmap.org/pipermail/osrm-talk/2017-February/001396.html) question on ml?

Thanks!
